# AI Assignment （Titanic) 
<hr>

## Name: Woo Chuan Jie
## Matric No.: U1620920J


In [3]:
import sklearn

ImportError: DLL load failed: The specified procedure could not be found.

In [1]:
import re

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.model_selection import KFold

ImportError: DLL load failed: The specified procedure could not be found.

In [2]:
test = pd.read_csv('test.csv')
train = pd.read_csv('train.csv')

# Going into Part A
<hr>

### Understand the training and testing datasets that we provided. Make a table to describe the two datasets, including their feature dimension, number of samples, mean and variance of the values in each attribute.

In [3]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [4]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Getting the feature dimension, number of samples, mean and variance of the values in each attribute of the dataset provided

In [5]:
test.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

In [6]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

# Going into Part B
<hr>

### Select at least one appropriate model (e.g., neural network, support vector machine, etc.) to build your classifier. Clearly describe the model you use, including the input and output dimensions, structure of the model, loss function(s), training strategy, etc. Include your code as well if you are solving the problem by programming.

<body>
I will be using logistic regression by python to create the classifier. Logistic regression desribe the possible outcomes of a single trial are moddled using logistic function. By inputing the data of the Pclass, gender and age i will be able to get the output of whether the people on board survive the tragedy or not. 

# Going into Part C
<hr>

### Discuss how you consider and determine the parameters (e.g., learning rate, etc.) / settings of your model as well as your reasons of doing so.

I will determine which parameters or settings by getting the accuracy of different parameters and choose the best out of them

# Going into Part D
<hr>

### Discuss how you handle attributes with missing values and what attributes are excluded from your classifier(s), and why.

Since I am using age as on of the main input, for null input of age, i will call them missing as their data are lost and not able to retrive. Since Ticket number and passenger fare is determined by the Pclass, embarked area, i will not included both the ticket number and passenger fare inside my parameter. I will not include data cabin number as I believe it does not have a huge impact on the survival rate of the people. 

In [7]:
def process_age(df,cut_points,label_names):
    df["Age"] = df["Age"].fillna(-0.5)
    df["Age_categories"] = pd.cut(df["Age"],cut_points,labels=label_names)
    return df

cut_points = [-1,0, 5, 12, 18, 35, 60, 100]
label_names = ["Missing", 'Infant', "Child", 'Teenager', "Young Adult", 'Adult', 'Senior']

train = process_age(train,cut_points,label_names)
test = process_age(test,cut_points,label_names)


In [8]:
def create_columns(df,column_name):
    new_column = pd.get_dummies(df[column_name],prefix=column_name)
    df = pd.concat([df,new_column],axis=1)
    return df

train = create_columns(train,"Pclass")
test = create_columns(test,"Pclass")
train = create_columns(train,"Sex")
test = create_columns(test,"Sex")
train = create_columns(train,"Age_categories")
test = create_columns(test,"Age_categories")

In [9]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin  ... Pclass_3 Sex_female  \
0  34.5      0      0   330911   7.8292   NaN  ...        1          0   
1  47.0      1      0   363272   7.0000   NaN  ...        1          1   
2  62.0      0      0   240276   9.6875   NaN  ...        0          0   
3  27.0      0      0   315154   8.6625   NaN  ...        1          0   
4  22.0      1      1  3101298  12.2875   NaN  ...        1          1   

   Sex_male  Age_categories_Missing  Age_categories_Infant  \
0         1                       0                      0   
1         0                       0                      0   
2         1                       0                      0   
3         1                       0                      0   
4         0                       0                      0   

   Age_categories_Child  Age_categories_Teenager  Age_categories_Young Adult  \
0                     0                        0                           1   
1                     0                        0                           0   
2                     0                        0                           0   
3                     0                        0                           1   
4                     0                        0                           1   

   Age_categories_Adult  Age_categories_Senior  
0                     0                      0  
1                     1                      0  
2                     0                      1  
3                     0                      0  
4                     0                      0  

[5 rows x 24 columns]

In [10]:
columns = ['Pclass_1', 'Pclass_2', 'Pclass_3', 'Sex_female', 'Sex_male',
       'Age_categories_Missing','Age_categories_Infant',
       'Age_categories_Child', 'Age_categories_Teenager',
       'Age_categories_Young Adult', 'Age_categories_Adult',
       'Age_categories_Senior']

all_X = train[columns]
all_y = train['Survived']

train_X, test_X, train_y, test_y = train_test_split(all_X, all_y, test_size=0.20,random_state=0)

In [11]:
lr = LogisticRegression()
lr.fit(train_X, train_y)
predictions = lr.predict(test_X)
accuracy = accuracy_score(test_y, predictions)
accuracy

0.81005586592178769

In [12]:
conf_matrix = confusion_matrix(test_y, predictions)
pd.DataFrame(conf_matrix, columns=['Survived', 'Died'], index=[['Survived', 'Died']])

Survived  Died
Survived        96    14
Died            20    49

In [13]:
scores = cross_val_score(lr, all_X, all_y, cv=13)
np.mean(scores)

0.80360726230291446

In [14]:
lr = LogisticRegression()
lr.fit(all_X, all_y)
test_predictions = lr.predict(test[columns])

# Going into Part E
<hr>

### Apply the classifier(s) built to the test set (“test.csv”) which contains the data of 418 passengers not included in the training set. Submit the “submission.csv” with the results you obtained.

In [15]:
test_ids = test["PassengerId"]
submission_df = {"PassengerId": test_ids,
                 "Survived": test_predictions}
submission = pd.DataFrame(submission_df)

submission.to_csv('submission.csv', index=False)

# Going into Part F
<hr>

### How many of these 418 passengers are classified as survivors from the Titanic tragedy? Among these classified survivors, how many of them were (i) female, (ii) below 18 years old, (iii) without any other family members on board? Passengers from which (iv) ticket class and (v) port of embarkment had the least chance of surviving the tragedy?

In [9]:
# test2 = pd.read_csv('test.csv')
duplicate = test.copy()
duplicate = pd.concat([duplicate,submission['Survived']],axis=1, sort=False)
# submission
duplicate.head()

NameError: name 'test' is not defined

# Qn F) Finding number of survivors from the titanic tragedy

In [17]:
Survived = duplicate[duplicate['Survived'] == True]
len(Survived)

145

# Qn Fi) Finding number of people who survived are female

In [18]:
female = Survived[Survived['Sex'] == 'female']
len(female)

144

# Qn Fii) Finding number of people who survived are age 18 and below

In [19]:
below_18 = Survived[Survived['Age'] < 18]
len(below_18)

43

# Qn Fiii) Finding number of people who survived without any other family members onboard

In [20]:
No_Siblings = Survived[Survived['SibSp'] == 0]
No_Spourse = No_Siblings[No_Siblings['Parch'] == 0]
len(No_Spourse)

65

# Qn Fiv) Passengers from which ticket class  had the least chance of surviving the tragedy?

In [21]:
# P1 = len(duplicate[duplicate['Pclass_1'] == 1])
Total = len(duplicate)
P1 = Total
S1 = len(Survived[Survived['Pclass_1']== 1])
Percentage1=S1/P1 * 100
# P2 = len(duplicate[duplicate['Pclass_2'] == 1])
P2 = Total
S2 = len(Survived[Survived['Pclass_2']== 1])
Percentage2=S2/P2 * 100
# P3 = len(duplicate[duplicate['Pclass_3'] == 1])
P3 = Total
S3 = len(Survived[Survived['Pclass_3']== 1])
Percentage3=S3/P3 * 100
print('Percentage of survivors from Pclass 1 =',round(Percentage1,3),'%')
print('Percentage of survivors from Pclass 2 =',round(Percentage2,3),'%')
print('Percentage of survivors from Pclass 3 =',round(Percentage3,3),'%')
compare = [Percentage1,Percentage2,Percentage3]
least = compare[0]
number = 1
for i in compare:
    if least > i:
        least = i
for checking in compare:        
    if least != checking:
        number = number + 1
    else:
        break
        
print('Passenger with ticket class',number,'had the least chance to survive the tragedy')

Percentage of survivors from Pclass 1 = 11.962 %
Percentage of survivors from Pclass 2 = 7.416 %
Percentage of survivors from Pclass 3 = 15.311 %
Passenger with ticket class 2 had the least chance to survive the tragedy


# Qn Fv) Passengers from which port of embarkment had the least chance of surviving the tragedy?

In [22]:
# EQ = len(duplicate[duplicate['Embarked'] == 'Q'])
EQ = Total
SQ = len(Survived[Survived['Embarked']== 'Q'])
PQ = SQ/EQ * 100
# EC = len(duplicate[duplicate['Embarked'] == 'C'])
EC = Total
SC = len(Survived[Survived['Embarked']== 'C'])
PC = SC/EC * 100
# ES = len(duplicate[duplicate['Embarked'] == 'S'])
ES=Total
SS = len(Survived[Survived['Embarked']== 'S'])
PS = SS/ES * 100
print('Percentage of people embarked from Queentown have survived =',round(PQ,3),'%')
print('Percentage of survivors embarked from Cherbourg =',round(PC,3),'%')
print('Percentage of survivors embarked from Southampton =',round(PS,3),'%')

compare = [PQ,PC,PS]
least = compare[0]
number = 1
for i in compare:
    if least > i:
        least = i
for checking in compare:        
    if least != checking:
        number = number + 1
    else:
        break
if number == 1:
    port = 'Queentown'
if number == 2:
    port = 'Cherbourg'
if number == 3:
    port = 'Southampton'  
    

print('Passenger who embarked from',port,'had the least chance to survive the tragedy')

Percentage of people embarked from Queentown have survived = 5.502 %
Percentage of survivors embarked from Cherbourg = 9.091 %
Percentage of survivors embarked from Southampton = 20.096 %
Passenger who embarked from Queentown had the least chance to survive the tragedy


# Going into Part G
<hr>

### Build one more classifier. Discuss and compare the results obtained from different classifiers.

By using support vector machine(svm), it gives a lower accuracy(78%)(as shown below) as compared when i using logistic regression(81%). Even thought svm is effective in high dimensional spaces and uses a subset of training points in the decision function so it is also memory efficient, but svm does not provide probability estimates.

In [23]:
svm = SVC(kernel = "linear",C=0.025,random_state=101)
svm.fit(train_X, train_y)
predictions_2 = svm.predict(test_X)
accuracy = accuracy_score(test_y, predictions_2)
accuracy

0.78770949720670391